tidb/src/main/java/io/tidb/bigdata/tidb/DeduplicateRowBuffer.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2022 TiDB Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.tidb.bigdata.tidb;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import io.tidb.bigdata.tidb.meta.TiIndexColumn;
+import io.tidb.bigdata.tidb.meta.TiIndexInfo;
+import io.tidb.bigdata.tidb.meta.TiTableInfo;
+import io.tidb.bigdata.tidb.row.Row;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class DeduplicateRowBuffer extends RowBuffer {
+
+  private final List<TiIndexInfo> uniqueIndexs;
+  private final RowUKBiMap rowUKBiMap;
+
+  public DeduplicateRowBuffer(
+      TiTableInfo tiTableInfo, boolean ignoreAutoincrementColumn, int bufferSize) {
+    // just use set to accelerate remove
+    super(bufferSize, new LinkedHashSet<>());
+    uniqueIndexs = SqlUtils.getUniqueIndexes(tiTableInfo, ignoreAutoincrementColumn);
+    rowUKBiMap = new RowUKBiMap();
+  }
+
+  // return true if there is no conflict
+  @Override
+  public boolean add(Row row) {
+    if (isFull()) {
+      throw new IllegalStateException("Row buffer is full!");
+    }
+    if (uniqueIndexs.size() == 0) {
+      rows.add(row);
+      return true;
+    }
+    boolean conflict = false;
+    for (TiIndexInfo indexInfo : uniqueIndexs) {
+      // get uniqueKeyColumns
+      List<TiIndexColumn> indexColumns = indexInfo.getIndexColumns();
+      List<Object> indexValue =
+          Collections.unmodifiableList(
+              indexColumns.stream()
+                  .map(i -> row.get(i.getOffset(), null))
+                  .collect(Collectors.toList()));
+      UniqueKeyColumns uniqueKeyColumns = new UniqueKeyColumns(indexInfo, indexValue);
+      // judge if conflict with old row
+      if (rowUKBiMap.uniqueKeyColumnsExist(uniqueKeyColumns)) {
```

Comment:
can we use `conflict = rowUKBiMap.get(xxx)==null` like:
```
      Row oldRow = rowUKBiMap.getRow(uniqueKeyColumns);
      if (oldRow != null) {
        conflict = true;
        rows.remove(oldRow);
        rowUKBiMap.removeRow(oldRow);
      }
```

Reply:
HashMap's value can be null, though it may not happen in our code.
thus, I think containskey may  be better